Add URL handling to chat

FreezyLemon opened the pull request:
Closes #806.

- [x] Depends on https://github.com/ppy/osu-framework/pull/1273

The old `LinkFormatter` class was mostly copied over and renamed to `MessageFormatter`.
It's implemented to format messages on `ChatLine.Message`'s setter.

For the actual link sprite, I took inspiration in the `OsuLinkSpriteText` class that was in ProfileHeader. I extended it and let `ChatLink` inherit it, which does all the stuff regarding click events and tooltips in chat.
The click events are not complete since not everything is implemented yet. So at the moment, some links will give a notification saying exactly that (see `OsuGame`).
I added a new `GetBeatmapRequest` which gets a map by map ID (not set ID), so that I could get the set ID off of that.

I also added a new test case and fixed the old `ChatDisplay` test case since it would break without the necessary dependencies.
The new interface `ICanDisableHoverSounds` adds a bool which lets a Parent element decide on whether to play the hover sound or not. This is needed so that the hover sound is not played too often (every word is a sprite, so if a link is multiple words long it would play the hover sound every time you move between words). If the interface is not implemented however, it is just ignored and the sound is played regardless (this is done for ease of use and not to break existing implementations).

The change in `PreviewButton` is something I found in testing: Sometimes preview.Length can be 0 (i don't quite remember why this happens), and when it is, it breaks its own Width (NaN) and crashes.

FreezyLemon committed:
Re-implemented message formatting (mostly taken from osu-stable code)

FreezyLemon committed:
Rolled back the idea that there should be a separate class for formatted messages

FreezyLemon committed:
Moved LinkFlowContainer out of ProfileHeader to make it available for other uses too (e.g. chat) and renamed it to LinkTextFlowContainer bc it can contain both links and text, not only one

FreezyLemon committed:
Also moved LinkText to its own file so the chat could reuse it (ProfileHeader's private class ProfileLink also still inherits from this, though)

FreezyLemon committed:
Added new class for chat lines, that colour the messages after formatting. URLs will become blue, and on hover (also defined here) be turned yellow-ish

FreezyLemon committed:
Fixed casting

FreezyLemon committed:
implemented method formatting into chat. Also added all necessary files to the .csproj

FreezyLemon committed:
Merge branch 'master' of https://github.com/ppy/osu into url-parsing-support

FreezyLemon committed:
Updated implementation to be based around a "LinkId" (atm the position of the link, anything unique to a link inside its message will be fine), which does not allow matching (OnHover related) between different links

FreezyLemon committed:
Added new request for getting the beatmapset from only a difficulty map ID

FreezyLemon committed:
Changed the LinkID to the link's own Index instead of the previous one (just makes more sense imo)

FreezyLemon committed:
Added all files to the .csproj and also introduced basic action filtering when you set the URL on an OsuLinkSpriteText object

FreezyLemon committed:
Fixed a bug where Drawable.Width could potentially be set to NaN (0/0) by checking if last variable > 0.

FreezyLemon committed:
Made the Chat testcase include a beatmapsetoverlay so links can be clicked from in there. Also had to implement private DI to make it work

FreezyLemon committed:
Merge branch 'url-parsing-support' of https://github.com/freezylemon/osu into url-parsing-support

FreezyLemon committed:
Changed URL detection to be more reliable and generally work better

FreezyLemon committed:
Removed unnecessary "using" statements

FreezyLemon committed:
Added ability to open other channels from chat links.

FreezyLemon committed:
Made the style a bit better, fixed a bug or two

FreezyLemon committed:
LoadComponent before assigning URL to ensure dependency loading

FreezyLemon committed:
Added osu.ppy.sh/ss (screenshot) URL handling and calculate ID at assignment so it's not on click anymore

FreezyLemon committed:
Added chatoverlay caching to testcase so test still works (chat needs to be injected so channels can be opened)

FreezyLemon committed:
Added "getIdFromUrl" call back to content.action because performance impact is small and no unnecessary id calculations are done

FreezyLemon committed:
Added new OsuLinkSpriteText.TextColour property that sets the internal content (OsuHoverContainer)'s colour instead of the whole container, so that text colour is always changed through that (e.g. link colouring, link hover fade).

Implemented it to be used when adding text to an OsuLinkTextFlowContainer.

FreezyLemon committed:
Made default link ID -1 (if no link is present) because linkId is currently being set to link.index which can be 0.

FreezyLemon committed:
Made it so the link ID is always added before loading the SpriteTexts (fixed weird bug where some sprites would be white instead of blue). Also improved XML doc on TextColour

FreezyLemon committed:
Added licence header and removed unnecessary whitespace

FreezyLemon committed:
Fixed messages sent by yourself not being formatted

FreezyLemon committed:
Added basic "on click" actions to the in-chat links

FreezyLemon committed:
Added centralised handling for some chat links

FreezyLemon committed:
Changed it so the ":" character does not appear after usernames when the displayed message is an action (e.g. /np). Just makes more sense imo

FreezyLemon committed:
Renamed ChatLinkSpriteText to ChatLink for convenience

FreezyLemon committed:
Added new test case for URL parsing / link display in chat

FreezyLemon committed:
Fixed up style in the ChatDisplay test

FreezyLemon committed:
Permitted nulls in ChatOverlay.load() so that in testing, no "OsuGame" instance is required. Also added null checks to the links' on click actions

FreezyLemon committed:
Fixed a bug where links would be out of order in their List which would cause the game to crash

FreezyLemon committed:
Various test fixes

FreezyLemon committed:
Removed unnecessary copy (pass-by-reference anyways)

FreezyLemon committed:
Added tooltip to links

FreezyLemon committed:
Removed "wiki:" prefix from wiki links (links are visible on tooltips so this is unnecessary now)

FreezyLemon committed:
Implemented new interface which allows parent containers to decide on whether a "OnHover" sound should be played.

FreezyLemon committed:
Implemented the new IHasHoverSounds interface in a private "ChatHoverContainer" class which is now used for ChatLink instances.
Also moved the overhead for finding all sprites in the same line that reference the same URL to the LoadComplete (used to be every hover, now only once).

FreezyLemon committed:
Added licence header to the new test case

FreezyLemon committed:
Added some tooltip text handling.
Also fixed a bug caused by SameLinkSprites where "this" is actually supposed to be included by just adding a manual function call.

FreezyLemon committed:
Fixed a bug where the hover wouldn't work correctly due to OsuLinkSpriteText assigning Action to a private property.

FreezyLemon committed:
Renamed the "on click" method to OnClick() and moved most of the implementation to ChatLink.
Also fixed the tooltip text up

FreezyLemon committed:
Set an empty list as default for AvailableChannels (mostly so that tests don't break, but also so that if no connection exists, the links don't break)

FreezyLemon committed:
Moved "Does this channel exist" check to ChatLine so that if a #name does not exist as a channel, it does not get added as a link (and does not contain a URL or anything else to prevent jankiness)

FreezyLemon committed:
Removed TODO and added another test line

FreezyLemon committed:
Renamed OnClick to OnLinkClicked to (what should be) obvious reasons

FreezyLemon committed:
Removed unnecessary whitespace

FreezyLemon committed:
Small style changes

FreezyLemon committed:
Reverted GetUserRequest because spectating is not implemented yet, and thus the additions are not needed (yet)

FreezyLemon committed:
*ahem* REVERTED the changes to GetUserRequest because they're not needed (yet)

FreezyLemon committed:
Removed unnecessary "using" directives

FreezyLemon committed:
Added parameter for LoadMultiplayerLobby (CI warning for unused int)

FreezyLemon committed:
Removed unused property and unnecessary return statement (CI)

FreezyLemon committed:
various CI adjustments

FreezyLemon committed:
Renamed the static variables and made them readonly, aswell as other small adjustments (CI)

FreezyLemon committed:
Renamed the IHasHoverSounds interface to ICanDisableHoverSounds and made it so that if the interface is not implemented, it is just ignored (samples will always be played). If it is implemented, the ShouldPlayHoverSound bool is decisive of whether sounds are played or not

FreezyLemon committed:
Trimmed whitespace (CI)

peppy commented on osu.Game/Graphics/UserInterface/HoverSounds.cs:
```diff
@@ -50,4 +51,13 @@ public enum HoverSampleSet
+    /// Classes implementing this interface can choose whether or not the HoverSounds should be played.
+    /// <para>If this is not implemented, the sounds will always be played when OnHover is triggered.</para>
+    /// </summary>
+    public interface ICanDisableHoverSounds
```
you should find a way to do this without doing this

peppy commented on osu.Game/Online/Chat/ChatLink.cs:
```diff
@@ -0,0 +1,229 @@
+
+            if (url.StartsWith("osu://"))
+            {
+                url = url.Substring(6);
```
please no magic substrings

peppy commented on osu.Game/Online/Chat/ChatLink.cs:
```diff
@@ -0,0 +1,229 @@
+
+        private int getIdFromUrl(string url)
+        {
+            var lastSlashIndex = url.LastIndexOf('/');
```
this should probably be regex

FreezyLemon committed:
Merged https://github.com/ppy/osu master branch

FreezyLemon committed:
Fixed small merging mistake

FreezyLemon committed:
Removed ICanDisableHoverSounds interface

FreezyLemon committed:
Changed it so ChatLinks handle hover and click sounds themselves

FreezyLemon committed:
Removed private OsuHoverContainer "content" from OsuLinkSpriteText for more customization.

FreezyLemon committed:
Removed unnecessary container and fixed "HandleInput" for ChatLinks

FreezyLemon committed:
Let regex handle "getIdFromUrl"

FreezyLemon committed:
Switched to regex to handle most of the decision on what to do on click. Also updated the getIdFromUrl method to adapt to the changes

FreezyLemon committed:
Removed unnecessary usings.

peppy commented on osu.Game/Graphics/Sprites/OsuLinkSpriteText.cs:
```diff
@@ -0,0 +1,45 @@
+        protected override bool OnClick(InputState state)
+        {
+            OnLinkClicked();
+            return base.OnClick(state);
```
shouldn't this return true?

peppy commented on osu.Game/Overlays/BeatmapSet/PreviewButton.cs:
```diff
@@ -80,7 +80,7 @@ protected override void Update()
             base.Update();
 
-            if (Playing.Value && preview != null)
+            if (Playing.Value && preview != null && preview.Length > 0)
```
what is this doing in this PR?

FreezyLemon replied:
Whoops, that was a makeshift fix for the Direct panels crashing (has already been resolved in another PR as far as I'm aware). Will remove in a second

FreezyLemon committed:
Removed makeshift fix that has already been addressed elsewhere.

FreezyLemon committed:
Made the Link's OnClick() Method return true

peppy committed:
Merge branch 'master' into url-parsing-support

peppy committed:
Move hover sound/container to base implementation

peppy commented:
Adding dependency on https://github.com/ppy/osu-framework/pull/1273.

FreezyLemon committed:
Merge branch 'master' into url-parsing-support

FreezyLemon committed:
Fixed problems introduced by the merge

FreezyLemon commented:
As discussed, I changed the following things:
- Removed the `LinkId`
- Moved all logic à la "is this a link or not" out of `ChatLink` (which also means no parent accessors)
- Removed word wrapping by replacing spaces with `_` to avoid previous issues with hovering

Other changes:
- Changed to proper `OsuHoverContainer` usage
  - use `OsuHoverContainer.Action` to handle clicks
  - change _its_ colour instead of `ChatLink`'s **and** implemented a `OsuLinkSpriteText.TextColour` property that gets/sets `Content.Colour` -> `OsuHoverContainer.Colour` (see to-do)
  - changed `OsuHoverContainer` to fade back to the colour that is set on load instead of always white.
- moved `ChatLink.getId` into `OnLinkClicked` (as a local function) since it's only used in that method

Still to do:
- [ ] Implement word wrapping
- [ ] Move the `TextColour` property to `SpriteText`? This seems useful to avoid errors where `Content != this` and `Content` changes its colour, like here. This would usually overlap the two colours instead of properly changing.
  * Remove the cast inside `ChatLine.updateMessageContent`

FreezyLemon committed:
Removed LinkId and word wrapping (for now).
Also reimplemented the OsuHoverContainer properly

FreezyLemon committed:
CI adjustments

FreezyLemon committed:
further CI adjustments

FreezyLemon committed:
Fixed only the links being italics for /me messages

FreezyLemon committed:
Better style; removed initial sprites

FreezyLemon committed:
handle chat == null gracefully and remove link from message if channel not found

FreezyLemon committed:
Fixed links being removed too early

FreezyLemon committed:
Added more and fixed the old automated tests

FreezyLemon committed:
Refactor and general tidying up

FreezyLemon committed:
small style fixes plus new assert in test

FreezyLemon committed:
fixed constant naming

peppy commented on osu.Game/Online/Chat/ChatLink.cs:
```diff
@@ -0,0 +1,160 @@
+                                chat.OpenChannel(foundChannel ?? throw new ArgumentException($"Unknown channel name ({args[1]})."));
+                                break;
+                            case "edit":
+                                chat.Game?.LoadEditorTimestamp();
```
You should not be accessing game through chat. Please DI game if you need game.

Also you'll want to make the BDL method nullable as game won't always be available.

peppy commented on osu.Game/Online/Chat/ChatLink.cs:
```diff
@@ -0,0 +1,160 @@
+
+                                break;
+                            case "spectate":
+                                GetUserRequest req;
```
i really wouldn't bother adding the `GetUserRequest` stuff for now. it's not achieving anything.

peppy commented on osu.Game/Online/Chat/ChatLink.cs:
```diff
@@ -0,0 +1,160 @@
+
+                                break;
+                            default:
+                                throw new ArgumentException($"Unknown osu:// link at {nameof(ChatLink)} ({urlMatch.Groups["content"].Value}).");
```
This will cause a hard crash. Not sure an exception is what you're looking for here.

FreezyLemon replied:
Oh, right. For some reason I thought that "invalid" osu:// links would be filtered out already, my bad. Just ignoring them should be fine then.

peppy commented on osu.Game/Online/Chat/ChatLink.cs:
```diff
@@ -0,0 +1,160 @@
+            else
+                base.OnLinkClicked();
+
+            int getId(string input)
```
not a huge fan of this method. why not just `.TrimStart('#')`?

peppy commented on osu.Game/Online/Chat/ChatLink.cs:
```diff
@@ -0,0 +1,160 @@
+                if (Url == Text)
+                    return null;
+
+                var urlMatch = getUrlMatch();
```
really ugly that we have to do this here. going forward we will need to encapsulate the logic for each link type somehow, so it's not spread across multiple methods like this.

peppy commented on osu.Game/Online/Chat/MessageFormatter.cs:
```diff
@@ -0,0 +1,171 @@
+            }
+        }
+
+        public class Link : IComparable<Link>
```
Let's move this out so it's not a nested class.

peppy commented on osu.Game/OsuGame.cs:
```diff
@@ -115,6 +115,34 @@ private void load(FrameworkConfigManager frameworkConfig)
+            });
+        }
+
+        internal void LoadMultiplayerLobby(int lobbyId)
```
Should be `LoadMatch(matchId)` if anything (the lobby is a place you find many matches)

peppy commented on osu.Game/Overlays/BeatmapSetOverlay.cs:
```diff
@@ -139,6 +139,20 @@ protected override bool OnClick(InputState state)
             return true;
         }
 
+        public void ShowBeatmap(int beatmapId)
```
this should probably be done in a separate pull request (and should have the user arrive to the correct difficulty's screen, rather than ignoring that). implementing methods like this without a TODO or an associated issue is carelessly creating bugs due to lacking implementation.

FreezyLemon replied:
Sorry. Got a bit too excited there and just put it in carelessly.

peppy commented on osu.Game/Graphics/Containers/OsuLinkFlowContainer.cs:
```diff
@@ -0,0 +1,59 @@
+        /// <summary>
+        /// The colour for text (links override this). Will only be used for new text elements.
+        /// </summary>
+        public ColourInfo TextColour = Color4.White;
```
why is this required?

FreezyLemon replied:
It's for messages with a "special" sender ('with background' in code). At the moment, the whole `contentFlow` changes colour which is not what we want since if we do that, the colours will overlap since the links will still be blue. For example: a special sender with yellow colour will have green-ish links with the old way of doing this. So the Sprites should be assigned the color directly instead of the whole container.

peppy commented on osu.Game/Graphics/Containers/OsuLinkFlowContainer.cs:
```diff
@@ -0,0 +1,59 @@
+            // TODO: Remove this and get word wrapping working
+            text = text.Replace(' ', '_');
+
+            return AddText(text, link =>
```
rather than calling `AddText` and making every single `SpriteText` be an `OsuLink`, why not just do the link construction here? Means you won't need all this casting, default colour logic, generic T classes etc.

FreezyLemon replied:
This exists so that the link in `ProfileHeader` still works. Changing this class to creating the links itself (and to remove generics) would require a way of differentiating between general osu links and the more specific chat links. It would be easy to just put two different methods there or inherit this from another class just for chat links, but I'm really not sure what the preferred solution for this would be.

FreezyLemon replied:
I like the idea, but creating the sprite inside this means I can't use the `base.Add...` methods since all of them create the SpriteTexts based on the input text. If I want to do this differently I'd have to change `TextFlowContainer.Add` to not throw an exception anymore, and that doesn't seem intended. But if I changed it so we would _not_ let this class inherit from `TextFlowContainer`, a lot of the code would probably have to be rewritten

FreezyLemon commented:
How does the following idea sound:
Instead of letting the `ChatLink` itself decide on what to do (when it's clicked), how about the `MessageFormatter` passing the `Link` an "Action" or "Type" or something (enum) with an argument (string) based on the URL and used regex? That way we would avoid having to run the same kind of pattern matching multiple times.

The sprite then only has to decide on an action based on an enum and then use the argument in said action.
For example: Action `LinkAction.OpenBeatmapSet` with the argument "93523". The `ChatLink` only has to parse the string to an int and can then pass the mapset ID to the beatmap overlay which can then open.

FreezyLemon committed:
Changed chat link implementation according to review

peppy committed:
Merge branch 'master' into url-parsing-support

peppy commented:
Something has regressed in layout: https://streamable.com/ouzbx

Notice the fill flow flashing weirdly.

FreezyLemon commented:
Okay, so:
The `ChatLink.Content` (an `OsuHoverContainer` instance) doesn't update its `DrawSize` correctly.
It seems that the actual `Size` is correct, but the DrawSize isn't invalidated until `loadComplete` automatically invalidates it again. It might also be something with wrong `RequiredParentSizeToFit` values of the children, but I'm not sure about that.

peppy committed:
Merge branch 'master' into url-parsing-support

peppy committed:
Merge branch 'master' into url-parsing-support

peppy committed:
Update licence headers

peppy committed:
Use a better method of link compilation

Adds word wrap back, simplifies a lot.

FreezyLemon committed:
Fixed bugs and added tests

FreezyLemon committed:
Nicer code to remove non-existent channels from links

FreezyLemon committed:
added custom hoverclicksounds so links make sounds on hover&click

peppy committed:
Merge branch 'master' into url-parsing-support

peppy commented:
Links don't seem to work any more :(

FreezyLemon commented:
What link didn't work for you? Seems like everything still works to me..
(Note that in tests, only external links have a visible effect since the DI'd game instance doesn't exist)

FreezyLemon commented:
- [ ] ppy/osu-framework#1324

That fixes the tooltips not working. But I still can't reproduce what you're saying, links are working for me.

FreezyLemon commented:
Oh, I see what you mean now. Investigating..

peppy committed:
Merge branch 'master' into url-parsing-support

peppy committed:
Update framework

peppy committed:
Update with upstream changes

peppy committed:
Update framework

peppy committed:
Add temporary browser handling of beatmap difficulty loading

until we have an api method to support this.

peppy committed:
Improve reference xmldoc

peppy committed:
Remove unnecessary region and make methods public

peppy committed:
Remove placeholder methods in OsuGame

Let's add these when they can actually be implemented.

peppy committed:
Add more xmldoc

peppy commented:
i still need to do a pass on code quality, but also found:

- osump:// links don't seem to work (and aren't tested).
- osu:// links don't seem to work (and aren't tested).
- typing an editor link gobbles the portion after the
![Uploading Kapture 2018-01-17 at 20.35.18.gif…]()

peppy committed:
Merge remote-tracking branch 'upstream/master' into url-parsing-support

# Conflicts:
#	osu.Game/osu.Game.csproj

peppy committed:
Merge remote-tracking branch 'upstream/master' into url-parsing-support

peppy committed:
Remove usage of ValueTuple to allow for dynamic recompilation

peppy committed:
Add new (failing) tests

peppy committed:
Add back old non-visual tests

peppy committed:
Prepare tests for fixing

peppy committed:
Add many more tests

peppy committed:
Add support for markdown style links

peppy committed:
Fix osump links

peppy committed:
Remove unnecessary paren handling from regex

Can't find a reason for this to exist

peppy committed:
Add better channel test cases (testing non-existent channels)

peppy committed:
Fix text getting truncated after last link in chat line

peppy committed:
Merge branch 'master' into url-parsing-support

peppy commented:
All aforementioned issues are now fixed. I'm going to do a second pass over the code as I'm still not confident about some of what is going on here.

Safe to say it's in a working and tested state now, though.

peppy committed:
Improve clarity and coverage of unit tests

peppy committed:
Merge branch 'master' into url-parsing-support

peppy commented:
I think I'm happy with this one. @smoogipoo would you like the opportunity to review this before merging? I rewrote a good portion so if you feel the need to review then let me know and I'll wait for that to happen.

smoogipoo commented:
Yes, I would like to review this.

smoogipoo committed:
Merge branch 'master' into url-parsing-support
